src/raft.rs

```diff
@@ -1851,20 +1851,20 @@ impl<T: Storage> Raft<T> {
 
         // Ignore redundant inserts.
         // TODO: Remove these and have this function and related functions return errors.
-        if let Some(progress) = self.prs().get(id) {
-            // If progress.is_learner == learner, then it's already inserted as what it should be, return early to avoid error.
-            if progress.is_learner == learner {
+        if self.prs().get(id).is_some() {
+            // If progress is a learner, then it's already inserted as what it should be, return early to avoid error.
+            if self.prs().learner_ids().contains(&id) == learner {
                 info!("{} Ignoring redundant insert of ID {}.", self.tag, id);
                 return;
             }
-            // If progress.is_learner == false, and learner == true, then it's a demotion, return early to avoid an error.
-            if !progress.is_learner && learner {
+            // If progress is a voter, and learner == true, then it's a demotion, return early to avoid an error.
+            if self.prs().voter_ids().contains(&id) && learner {
                 info!("{} Ignoring voter demotion of ID {}.", self.tag, id);
                 return;
             }
         };
 
-        let progress = Progress::new(self.raft_log.last_index() + 1, self.max_inflight, learner);
+        let progress = Progress::new(self.raft_log.last_index() + 1, self.max_inflight);
```

Comment:
Better move it inside condition body. There will be an allocation when creating a progress.